merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/engine/SimulationEngine.java

```diff
@@ -390,13 +354,51 @@ public boolean isTaskComplete(final TaskId task) {
   // TODO: Whatever mechanism replaces `computeResults` also ought to replace `isTaskComplete`.
   // TODO: Produce results for all tasks, not just those that have completed.
   //   Planners need to be aware of failed or unfinished tasks.
-  public SimulationResults computeResults(
+  public static SimulationResults computeResults(
       final SimulationEngine engine,
       final Instant startTime,
       final Duration elapsedTime,
-      final Map<String, ActivityInstanceId> taskToPlannedDirective,
+      final Topic<ActivityInstanceId> activityTopic,
       final TemporalEventSource timeline,
       final MissionModel<?> missionModel) {
+    final Map<String, ActivityInstanceId> taskToPlannedDirective = new HashMap<>();
+
+    for (final var point : timeline) {
+      if (!(point instanceof TemporalEventSource.TimePoint.Commit p)) continue;
+      if (!p.topics().contains(activityTopic)) continue;
+
+      p.events()
+       // Filter events that match the given topic.
+       .substitute($ -> ($.topic() == activityTopic) ? EventGraph.atom($) : EventGraph.empty())
+       // Interpret each event as an action adding the event and its originating task to a collection.
+       .evaluate(
+           new EffectTrait<Consumer<Map<String, ActivityInstanceId>>>() {
+             @Override
+             public Consumer<Map<String, ActivityInstanceId>> empty() {
+               return $ -> {};
+             }
+
+             @Override
+             public Consumer<Map<String, ActivityInstanceId>> sequentially(
+                 final Consumer<Map<String, ActivityInstanceId>> prefix,
+                 final Consumer<Map<String, ActivityInstanceId>> suffix)
+             {
+               return $ -> { prefix.accept($); suffix.accept($); };
+             }
+
+             @Override
+             public Consumer<Map<String, ActivityInstanceId>> concurrently(
+                 final Consumer<Map<String, ActivityInstanceId>> left,
+                 final Consumer<Map<String, ActivityInstanceId>> right)
+             {
+               return $ -> { left.accept($); right.accept($); };
+             }
+           },
+           ev -> $ -> ev.extract(activityTopic).ifPresent(id -> $.put(ev.provenance().id(), id)))
```

Comment:
Some thoughts here - I renamed `id` with `directiveId` when reading this code because I got worried for a minute that our Map was backwards.

An invariant here is that no task will ever emit two different directive IDs to activityTopic. That seems reasonable, but I figured I'd call it out anyhow.

Reply:
Yes, and honestly (see Future Work), I'd like to see this handling of directive associations and input/output attributes pushed out past the Merlin API boundary altogether. A lot of the results code is basically an upstreamed analysis of the simulation, which ought to be done elsewhere.

Reply:
> 🔵 Some thoughts here - I renamed `id` with `directiveId` when reading this code because I got worried for a minute that our Map was backwards.

Good call; I've renamed the variable.
```diff
-          ev.extract(this.activityTopic).ifPresent(id -> taskInfo.taskToPlannedDirective.put(ev.provenance().id(), id));
+          ev.extract(this.activityTopic)
+            .ifPresent(directiveId -> taskInfo.taskToPlannedDirective.put(ev.provenance().id(), directiveId));
```